src/terminal_chat_console.cpp

```diff
@@ -0,0 +1,432 @@
+/*
+Minetest
+Copyright (C) 2015 est31 <MTest31@outlook.com>
+
+This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation; either version 2.1 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#include "config.h"
+#if USE_CURSES
+#include "version.h"
+#include "terminal_chat_console.h"
+#include "porting.h"
+#include "settings.h"
+#include "util/numeric.h"
+#include "util/string.h"
+
+// include this last to avoid any conflicts
+// (likes to set macros to common names, conflicting various stuff)
+#include "ncursesw/curses.h"
+
+// Some functions to make drawing etc position independent
+static void reformat_backend(ChatBackend *backend, int rows, int cols)
+{
+	backend->reformat(cols, rows - 2);
+}
+
+static void move_for_backend(int row, int col)
+{
+	move(row + 1, col);
+}
+
+void TerminalChatConsole::init_of_curses()
+{
+	initscr();
+	cbreak(); //raw();
+	noecho();
+	keypad(stdscr, TRUE);
+	nodelay(stdscr, TRUE);
+	timeout(100);
+
+	// To make esc not delay up to one second. According to the internet,
+	// this is the value vim uses, too.
+	set_escdelay(25);
+
+	getmaxyx(stdscr, m_rows, m_cols);
+	reformat_backend(&m_chat_backend, m_rows, m_cols);
+}
+
+void TerminalChatConsole::deinit_of_curses()
+{
+	endwin();
+}
+
+void TerminalChatConsole::resize()
+{
+	m_chat_backend.reformat(m_cols, m_rows - 2);
+}
+
+void *TerminalChatConsole::run()
+{
+	BEGIN_DEBUG_EXCEPTION_HANDLER
+
+	std::cout << "========================" << std::endl;
+	std::cout << "Begin log output over terminal"
+		<< " (no stdout/stderr backlog during that)" << std::endl;
+	// Make the loggers to stdout/stderr shut up.
+	// Go over our own loggers instead.
+	LogLevelMask err_mask = g_logger.removeOutput(&stderr_output);
+	LogLevelMask out_mask = g_logger.removeOutput(&stdout_output);
+
+	g_logger.addOutput(&m_log_output);
+
+	// Inform the server of our nick
+	m_chat_interface->command_queue.push_back(
+		ChatEvent(CET_NICK_ADD, m_nick, L""));
+
+	init_of_curses();
+
+	while (!stopRequested()) {
+
+		int ch = getch();
+		if (stopRequested())
+			break;
+
+		step(ch);
```

Comment:
Unless I'm misunderstanding something, step() combines both input processing AND queue processing.
This is bad because queues are only updated when input is received.  Doesn't this mean the administrator cannot passively watch the terminal console, since it'd be blocking on keyboard input instead of updating the screen?